tests/module/test_ipddp.py

```diff
@@ -0,0 +1,71 @@
+        super(InvPend, self).__init__()
+        self.tau = dt
+        self.length = length
+        self.gravity = gravity
+
+    def state_transition(self, state, input, t=None):
+        force = input.squeeze(-1)
+        _dstate = torch.stack([state[...,1], force+self.gravity/self.length[0]*torch.sin(state[...,0].clone())], dim=-1)
+        return state + torch.mul(_dstate, self.tau)
+
+    def observation(self, state, input, t=None):
+        return state
+
+if __name__ == "__main__":
+    torch.set_default_dtype(torch.float64)
+    # Create parameters for inv pendulum trajectory
+    dt = 0.05   # Delta t
+    T = 5    # Number of time steps
+
+    # Initial state
+    state = torch.tensor([
+            [-2.,0.],
+            [-1., 0.],
+            [-2.5, 1.]],
+            device=device)
+
+    sys = InvPend(dt)
+    ns, nc = 2, 1
+    n_batch = 3
+    state_all = torch.zeros(n_batch, T + 1, ns, device=device)
+    input_all = torch.ones(n_batch, T, nc, device=device) * 0.02
+    state_all[...,0,:] = state
+    init_traj = {'state': state_all,
+                 'input': input_all}
+
+    # Create cost object
+    Q = torch.tile(dt * torch.eye(ns, ns, device=device), (n_batch, T, 1, 1))
+    R = torch.tile(dt * torch.eye(nc, nc, device=device), (n_batch, T, 1, 1))
+    S = torch.tile(torch.zeros(ns, nc, device=device), (n_batch, T, 1, 1))
+    c = torch.tile(torch.zeros(1, device=device), (n_batch, T))
+    stage_cost = pp.module.QuadCost(Q, R, S, c)
+    terminal_cost = pp.module.QuadCost(10./dt*Q[...,0:1,:,:], R[...,0:1,:,:], S[...,0:1,:,:], c[...,0:1]) # special stagecost with T=1
+
+    # Create constraint object
+    gx = torch.tile(torch.zeros(2 * nc, ns, device=device), (n_batch, T, 1, 1))
+    gu = torch.tile(torch.vstack((torch.eye(nc, nc, device=device), - torch.eye(nc, nc, device=device)) ), (n_batch, T, 1, 1))
+    g = torch.tile(torch.hstack((- 0.25 * torch.ones(nc, device=device), - 0.25 * torch.ones(nc, device=device)) ), (n_batch, T, 1))
+    lincon = pp.module.LinCon(gx, gu, g)
+
+    traj_opt = [None for batch_id in range(n_batch)]
+
+    for batch_id in range(n_batch): # use for loop and keep the IPDDP
+        # used class is for batched version, inside loop use batch_size = 1
+        stage_cost = pp.module.QuadCost(Q[batch_id:batch_id+1], R[batch_id:batch_id+1], S[batch_id:batch_id+1], c[batch_id:batch_id+1])
+        terminal_cost = pp.module.QuadCost(10./dt*Q[batch_id:batch_id+1,0:1,:,:], R[batch_id:batch_id+1,0:1,:,:], S[batch_id:batch_id+1,0:1,:,:], c[batch_id:batch_id+1,0:1])
+        lincon = pp.module.LinCon(gx[batch_id:batch_id+1], gu[batch_id:batch_id+1], g[batch_id:batch_id+1])
+        init_traj_sample = {'state': init_traj['state'][batch_id:batch_id+1],
+                            'input': init_traj['input'][batch_id:batch_id+1]}
+        ipddp = IPDDP(sys, stage_cost, terminal_cost, lincon, gx.shape[-2], init_traj_sample)
+        traj_opt[batch_id] = ipddp.solver()
```

Comment:
Here, used solver, but in the training example, used forward.

Reply:
In the test, it only cares about solving the traj, and does not create a computational graph for learning; in the training example, it first calls solver() to solve the trajectory, then use forward to create the computational graph for autodiff.